extensions/openwsman/processors/SourceInitiatedSubscriptionListener.cpp

```diff
@@ -0,0 +1,918 @@
+  subscription_version_ = subscription_version;
+  subscription_ = subscription;
+  subscription_endpoint_ = subscription_endpoint;
+  subscription_identifier_ = subscription_identifier;
+}
+
+void SourceInitiatedSubscriptionListener::SubscriberData::clearSubscription() {
+  subscription_version_.clear();
+  if (subscription_ != nullptr) {
+    ws_xml_destroy_doc(subscription_);
+  }
+  subscription_ = nullptr;
+}
+
+void SourceInitiatedSubscriptionListener::SubscriberData::setBookmark(WsXmlDocH bookmark) {
+  clearBookmark();
+  bookmark_ = bookmark;
+}
+
+void SourceInitiatedSubscriptionListener::SubscriberData::clearBookmark() {
+  if (bookmark_ != nullptr) {
+    ws_xml_destroy_doc(bookmark_);
+  }
+  bookmark_ = nullptr;
+}
+
+/*
+ * This is very basic state storage. We will have to ensure that the configurations match before we use a bookmark,
+ * and we also have to remove serialized bookmarks when needed. Once the proper centralized state storage is merged back,
+ * it will be much easier to implement a proper state storage with that, so for the time being I will leave it like this.
+ */
+bool SourceInitiatedSubscriptionListener::persistState() const {
+  utils::file::FileUtils::create_dir(state_directory_path_);
+  for (const auto& subscriber : subscribers_) {
+    std::ofstream bookmark_file(utils::file::FileUtils::concat_path(state_directory_path_, subscriber.first));
+    char* xml_buf = nullptr;
+    int xml_buf_size = 0;
+    ws_xml_dump_memory_enc(subscriber.second.bookmark_, &xml_buf, &xml_buf_size, "UTF-8");
+    bookmark_file.write(xml_buf, xml_buf_size);
+    ws_xml_free_memory(xml_buf);
+  }
+  return true;
+}
+
+bool SourceInitiatedSubscriptionListener::loadState() {
+  std::map<std::string /*machineId*/, WsXmlDocH /*bookmark*/> bookmarks;
+  utils::file::FileUtils::list_dir(state_directory_path_, [&](const std::string& directory, const std::string& filename) -> bool {
+    WsXmlDocH bookmark = ws_xml_read_file(utils::file::FileUtils::concat_path(directory, filename).c_str(), "UTF-8", 0);
+
+    /*
+     * Unless we duplicate this doc here, it causes weird errors later, much like the later ws_xml_copy_node
+     * could not make a perfect deep copy. TODO(bakaid): investigate
+     */
+    WsXmlNodeH bookmark_node = ws_xml_get_doc_root(bookmark);
+    WsXmlDocH bookmark_doc = ws_xml_create_doc(XML_NS_WS_MAN, WSM_BOOKMARK);
+    WsXmlNodeH temp = ws_xml_get_doc_root(bookmark_doc);
+    ws_xml_duplicate_children(temp, bookmark_node);
+    ws_xml_destroy_doc(bookmark);
+
+    bookmarks[filename] = bookmark_doc;
+    return true;
+  }, logger_);
+
+  for (const auto& bookmark : bookmarks) {
+    subscribers_[bookmark.first].setBookmark(bookmark.second);
+  }
+  return true;
+}
+
+std::string SourceInitiatedSubscriptionListener::Handler::millisecondsToXsdDuration(int64_t milliseconds) {
+  char buf[1024];
+  snprintf(buf, sizeof(buf), "PT%lld.%03lldS", milliseconds / 1000, milliseconds % 1000);
+  return buf;
```

Comment:
And 3 bytes for the milliseconds. That's why I chose a sensibly large buffer instead of potentially miscalculating the required size, although I admit it is more than enough.